### DISCUSSION TITLE
Can I add a parameter when binding to a task result, if so how ?

### DISCUSSION THREAD

**Discussion by Yveske1966** [AvaloniaUI/Avalonia#14561] [Q&A] [answered]
Using https://docs.avaloniaui.net/docs/guides/data-binding/how-to-bind-to-a-task-result
This works, but I would like to call this from within a listbox to fetch the correct image on demand from a second table.

ImageCardControl.cs
![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/5e5ba586-809b-48ae-9b66-41c1f83829d6)

ImageCardControl.axaml
![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/da5a74f2-f297-499a-8613-7ae2eaf44c6f)

PicturesView.axaml
![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/26078de4-ba6d-490d-8f51-915e4e0155fc)

PicturesViewModel.cs
![image](https://github.com/AvaloniaUI/Avalonia/assets/68047793/8062efe0-ea1a-46ce-9af7-ee9ee2e2305d)

GetImageFromDatabaseAsync should return an image based upon a parameter or the control that called the function.
Is this possible and is my approach correct or should I use a different strategy ?

Many thanks in advance,
Yveske

**Comment by thevortexcloud** [comment]
Do you specifically need a parameter? If not the simplest thing I can think of is to make your load method reference some property on your VM.

>  or the control that called the function.

This sounds somewhat suspicious to me. Typically the specific control should not matter, only the data context. In any case, I am not sure if there is a direct way to do what you are asking.

**Reply by Yveske1966** [reply]
When the View starts, all info out of Image___List is listed and for the first couple of Items in the ListBox, the function GetImageFromDatabaseAsync is called. When I scroll (up/down in my case), the function is called again for the next couple of Items.

Adding the image to the Image___List would be the easiest solution, but that means all images are loaded on screen at startup, which takes forever and bombs out if the user makes the wrong request (eg. loading all +500,000 images) in the menu.

I'm trying to find a way to know which image is requested when the ListBox fills its cache, because there is no connection to the "current" item in the UI. There is also no connection to the visible items in the UI, as the ListBox loads into its cache a couple of Item images more than will fit on screen.

In the recording below you can see that even for 783 images in the ListBox, the function is only called 4 times (when not scrolling)
https://github.com/AvaloniaUI/Avalonia/assets/68047793/9b9c0bae-86a8-418f-9812-e3fb9b0ff482

**Reply by thevortexcloud** [reply, marked as answer]
> Adding the image to the Image___List would be the easiest solution, but that means all images are loaded on screen at startup, which takes forever and bombs out if the user makes the wrong request (eg. loading all +500,000 images) in the menu.

This really sounds like you want a virtualised control (which is something ListBox can do if you enable it by using a virtualised panel), which will handle all of this for you. In an app I have, I load >2k image VMs all at once into a single collection and bind to that using a virtualised panel. I then bind to a task to actually load the image when the control becomes realised. Performance is pretty good, and my memory usage (when loading from a disk cache) is typically ~200mb total. So the image does not exist until the user actually sees it.

> I'm trying to find a way to know which image is requested when the ListBox fills its cache

I don't know if you can do this with a ListBox. You can do it fairly easily with an ItemsRepeater though. It has an [event](https://learn.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.itemsrepeater.elementprepared?view=winui-2.2) (the API for ItemRepeater in Avalonia is identical to WinUIs) specifically for that. The downside with ItemsRepeater is it has no concept of selection out of the box. So you need to implement all that yourself. It's not too hard, but it's fairly annoying. I have done it myself reusing `ListBoxItem` as my item container.

In any case, it sounds like you are possibly trying to solve problems Avalonia has already solved.

On a completely unrelated note: I am curious as to why you are not using XML comments. A lot of .NET tooling (including Visual Studio) will read it to provide information about a property/method/variable/etc.

**Reply by Yveske1966** [reply]
Thanks for your answer, sounds like I have some (a lot of) studying to do.
Will let you know if I could solve my problem that way and mark this as an answer.

Will have a look at XML comments as well.
I'm coming from C++Builder and it's my first encounter with Visual Studio and C#. 
Lots of learning to do...

**Reply by thevortexcloud** [reply]
You are welcome! Let me know if you have a follow up question too. 

> I'm coming from C++Builder and it's my first encounter with Visual Studio and C#.

Fair enough, in that case I will also add that your [variable(?) names do not follow the standard naming conventions for C# ](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names). Which makes reading your code slightly harder for me. A lot of analysers/IDEs will actually raise this sort of stuff as an error/warning even if it compiles fine.

**Reply by thevortexcloud** [reply]
Oh also, here is the discussion I made where I asked about the best way to lazy load a lot of images.

https://github.com/AvaloniaUI/Avalonia/discussions/13605

**Reply by Yveske1966** [reply]
Thanks again. Will certainly have a look at it !

**Reply by Yveske1966** [reply]
Got it to work with ItemsRepeater, thanks !
